crates/phactory/src/system/side_tasks/geo_probe.rs

```diff
@@ -0,0 +1,181 @@
+pub enum GeoProbeError {
+    // geo_probe
+    FailedToGetPublicIPAddress,
+    DBNotFound,
+    DBNotValid,
+    IPNotValid,
+    NoRecord,
+    UnknownError,
+}
+
+impl fmt::Display for GeoProbeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GeoProbeError::FailedToGetPublicIPAddress => {
+                write!(f, "network error, failed to get public IP address")
+            }
+            GeoProbeError::DBNotFound => write!(f, "geolite DB not found"),
+            GeoProbeError::DBNotValid => write!(f, "geolite DB is probably broken"),
+            GeoProbeError::IPNotValid => write!(f, "fetched IP address not valid for parsing"),
+            GeoProbeError::NoRecord => write!(f, "no record found in DB"),
+            GeoProbeError::UnknownError => write!(f, "unknown error"),
+        }
+    }
+}
+
+impl error::Error for GeoProbeError {}
+
+pub fn db_query_region_name<'a>(city_general_data: &'a geoip2::City) -> Option<&'a str> {
+    if let Some(city) = &city_general_data.city {
+        return Some(*city.names.as_ref()?.get(&"en")?);
+    }
+
+    if let Some(subdivisions) = &city_general_data.subdivisions {
+        return Some(*subdivisions[0].names.as_ref()?.get(&"en")?);
+    }
+
+    if let Some(country) = &city_general_data.country {
+        return Some(*country.names.as_ref()?.get(&"en")?);
+    }
+
+    Some(&"N/A")
+}
+
+pub fn process_block(
+    block_number: BlockNumber,
+    egress: &Sr25519MessageChannel,
+    side_task_man: &mut SideTaskManager,
+    identity_key: &sr25519::Pair,
+    geoip_city_db: String,
+) {
+    let identity_key = identity_key.clone();
+    let worker_pubkey = identity_key.public();
+    let raw_pubkey: &[u8] = worker_pubkey.as_ref();
+    let pkh = blake2_256(raw_pubkey);
+    let (pkh_first_32_bits, _) = pkh.split_at(std::mem::size_of::<u32>());
+    let worker_magic = u32::from_be_bytes(
+        pkh_first_32_bits.try_into().expect("Should never fail with valid worker pubkey; qed."),
+    ) % BLOCK_INTERVAL;
+    if block_number % BLOCK_INTERVAL == worker_magic {
+        log::info!(
+            "start geolocation probing at block {}, worker magic {}",
+            block_number,
+            worker_magic
+        );
+
+        let egress = egress.clone();
+        let duration = PROBE_DURATION;
+        let task = AsyncSideTask::spawn(
+            block_number,
+            duration,
+            async {
+                // 1. we load the database first, so that in case where the database not exists,
+                // we can just return an error without emits any http request.
+                let geo_db_buf =
+                    std::fs::read(geoip_city_db).map_err(|_| GeoProbeError::DBNotFound)?;
+
+                // 2. get IP address.
+                let mut resp = surf::get(IP_PROBE_URL)
+                    .send()
+                    .await
+                    .map_err(|_| GeoProbeError::FailedToGetPublicIPAddress)?;
```

Comment:
When the error is a trivial emum variant, I'd prefer to use `.or(Err(...))` than `.map_err(|_| ...)`. The later method is perfect when it's non-trivial to build an error, for example, run format to produce some error message.